backend/zkppSalt.js

```diff
@@ -0,0 +1,229 @@
+  if (record) {
+    return false
+  }
+
+  const encryptionKey = hashStringArray('REG', contract, registrationSecret).slice(0, nacl.secretbox.keyLength)
+  const nonce = nacl.randomBytes(nacl.secretbox.nonceLength)
+  const keyPair = boxKeyPair()
+  const s = Buffer.concat([nonce, nacl.secretbox(keyPair.secretKey, nonce, encryptionKey)]).toString('base64').replace(/\//g, '_').replace(/\+/g, '-').replace(/=*$/, '')
+  const now = (Date.now() / 1000 | 0).toString(16)
+  const sig = [now, Buffer.from(hashStringArray(contract, b, s, now, challengeSecret)).toString('base64').replace(/\//g, '_').replace(/\+/g, '-').replace(/=*$/, '')].join(',')
+
+  return {
+    s,
+    p: Buffer.from(keyPair.publicKey).toString('base64').replace(/\//g, '_').replace(/\+/g, '-').replace(/=*$/, ''),
+    sig
+  }
+}
+
+export const register = async (contract: string, clientPublicKey: string, encryptedSecretKey: string, userSig: string, encryptedHashedPassword: string): Promise<boolean> => {
+  if (!verifyChallenge(contract, clientPublicKey, encryptedSecretKey, userSig)) {
+    return false
+  }
+
+  const record = await getZkppSaltRecord(contract)
+
+  if (record) {
+    return false
+  }
+
+  const encryptedSecretKeyBuf = Buffer.from(encryptedSecretKey.replace(/_/g, '/').replace(/-/g, '+'), 'base64')
+  const encryptionKey = hashStringArray('REG', contract, registrationSecret).slice(0, nacl.secretbox.keyLength)
+  const secretKeyBuf = nacl.secretbox.open(encryptedSecretKeyBuf.slice(nacl.secretbox.nonceLength), encryptedSecretKeyBuf.slice(0, nacl.secretbox.nonceLength), encryptionKey)
+
+  const parseRegisterSaltRes = parseRegisterSalt(clientPublicKey, secretKeyBuf, encryptedHashedPassword)
+
+  if (!parseRegisterSaltRes) {
+    return false
+  }
+
+  const [authSalt, contractSalt, hashedPasswordBuf] = parseRegisterSaltRes
+
+  await setZkppSaltRecord(contract, Buffer.from(hashedPasswordBuf).toString(), authSalt, contractSalt)
+
+  return true
+}
+
+const contractSaltVerifyC = (h: string, r: string, s: string, userHc: string) => {
+  const [c, hc] = computeCAndHc(r, s, h)
+  const userHcBuf = Buffer.from(userHc.replace(/_/g, '/').replace(/-/g, '+'), 'base64')
+
+  if (hc.byteLength === userHcBuf.byteLength && timingSafeEqual(hc, userHcBuf)) {
+    return c
+  }
+
+  return false
+}
+
+export const getContractSalt = async (contract: string, r: string, s: string, sig: string, hc: string): Promise<false | string> => {
+  if (!verifyChallenge(contract, r, s, sig)) {
+    return false
+  }
+
+  const record = await getZkppSaltRecord(contract)
+  if (!record) {
+    return false
+  }
+
+  const { hashedPassword, contractSalt } = record
+
+  const c = contractSaltVerifyC(hashedPassword, r, s, hc)
+
+  if (!c) {
+    return false
+  }
+
+  return encryptContractSalt(c, contractSalt)
+}
+
+export const update = async (contract: string, r: string, s: string, sig: string, hc: string, encryptedArgs: string): Promise<boolean> => {
```

Comment:
The name "update" is very vague and unclear. I have no idea what this function is doing or why. Could you please add commentary and improve the function name?

Reply:
Reminder